drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +323,476 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (data->async->rx_next_buf == NULL) {
+		data->async->rx_next_buf = buf;
+		nrf_uarte_rx_buffer_set(uarte, buf, len);
+	} else {
+		return -EBUSY;
+	}
+	return 0;
+}
+
+static int uarte_nrfx_callback_set(struct device *dev, uart_callback_t callback,
+				  void *user_data)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	data->async->user_callback = callback;
+	data->async->user_data = user_data;
+
+	return 0;
+}
+
+static int uarte_nrfx_rx_disable(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (data->async->rx_buf == NULL) {
+		return -EFAULT;
+	}
+
+	k_timer_stop(&data->async->rx_timeout_timer);
+	data->async->rx_enabled = false;
+
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPRX);
+
+	return 0;
+}
+
+static void tx_timeout(struct k_timer *timer)
+{
+	struct device *dev = k_timer_user_data_get(timer);
+	(void) uarte_nrfx_tx_abort(dev);
+}
+
+static void user_callback(struct device *dev, struct uart_event *evt)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	if (data->async->user_callback) {
+		data->async->user_callback(evt, data->async->user_data);
+	}
+}
+
+/**
+ * Whole timeout is divided by RX_TIMEOUT_DIV into smaller units, rx_timeout
+ * is executed periodically every rx_timeout_slab ms. If between executions
+ * data was received, then we start counting down time from start, if not, then
+ * we subtract rx_timeout_slab from rx_timeout_left.
+ * If rx_timeout_left is less than rx_timeout_slab it means that receiving has
+ * timed out and we should tell user about that.
+ */
+static void rx_timeout(struct k_timer *timer)
+{
+	struct device *dev = k_timer_user_data_get(timer);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	u32_t read;
+
+	if (hw_rx_counting_enabled(data)) {
+		read = nrfx_timer_capture(&data->async->timer, 0);
+	} else {
+		read = data->async->rx_cnt.cnt;
+	}
+
+	/* Check if data was received since last function call */
+	if (read != data->async->rx_total_byte_cnt) {
+		data->async->rx_total_byte_cnt = read;
+		data->async->rx_timeout_left = data->async->rx_timeout;
+	}
+
+	/* Check if there is data that was not sent to user yet */
+	if (data->async->rx_total_byte_cnt
+	    != data->async->rx_total_user_byte_cnt) {
+		if (data->async->rx_timeout_left
+		    < data->async->rx_timeout_slab) {
+			/* rx_timeout ms elapsed since last receiving */
+			u32_t len = data->async->rx_total_byte_cnt
+				    - data->async->rx_total_user_byte_cnt;
+			struct uart_event evt = {
+				.type = UART_RX_RDY,
+				.data.rx.buf = data->async->rx_buf,
+				.data.rx.len = len,
+				.data.rx.offset = data->async->rx_offset
+			};
+			data->async->rx_offset += len;
+			data->async->rx_total_user_byte_cnt =
+				data->async->rx_total_byte_cnt;
+			user_callback(dev, &evt);
+			data->async->rx_timeout_left = data->async->rx_timeout;
```

Comment:
This seems to be redundant. You can't enter this block again without first getting inside `if (read != data->async->rx_total_byte_cnt) {` above.